core/templates/dev/head/pages/story_editor/main_editor/StoryNodeEditorDirective.js

```diff
@@ -70,41 +115,94 @@ oppia.directive('storyNodeEditor', [
           };
 
-          $scope.addNewDestinationNode = function() {
-            var nextNodeId = $scope.story.getStoryContents().getNextNodeId();
-            StoryUpdateService.addStoryNode($scope.story);
-            StoryUpdateService.addDestinationNodeIdToNode(
-              $scope.story, $scope.getId(), nextNodeId);
-          };
-
-          $scope.addExistingDestinationNode = function(nodeId) {
-            StoryUpdateService.addDestinationNodeIdToNode(
-              $scope.story, $scope.getId(), nodeId);
-            $scope.newDestinationNodeId = '';
-            $scope.closeNodeIdEditor();
+          $scope.addDestinationNode = function(nodeId) {
+            if (nodeId === null) {
+              var nodeTitles =
+                $scope.story.getStoryContents().getNodes().map(function(node) {
+                  return node.getTitle();
+                });
+              var modalInstance = $uibModal.open({
+                templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+                  '/pages/story_editor/main_editor/' +
+                  'new_chapter_title_modal_directive.html'),
+                backdrop: true,
+                controller: [
+                  '$scope', '$uibModalInstance',
+                  function($scope, $uibModalInstance) {
+                    $scope.nodeTitle = '';
+                    $scope.nodeTitles = nodeTitles;
+                    $scope.errorMsg = null;
+
+                    $scope.resetErrorMsg = function() {
+                      $scope.errorMsg = null;
+                    };
+                    $scope.isNodeTitleEmpty = function(nodeTitle) {
+                      return (nodeTitle === '');
+                    };
+                    $scope.save = function(title) {
+                      if ($scope.nodeTitles.indexOf(title) !== -1) {
+                        $scope.errorMsg =
+                          'A chapter with this title already exists';
+                        return;
+                      }
+                      $uibModalInstance.close(title);
+                    };
+                    $scope.cancel = function() {
+                      $uibModalInstance.dismiss('cancel');
+                    };
+                  }
+                ]
+              });
+
+              modalInstance.result.then(function(title) {
+                var nextNodeId =
+                  $scope.story.getStoryContents().getNextNodeId();
+                StoryUpdateService.addStoryNode($scope.story, title);
+                StoryUpdateService.addDestinationNodeIdToNode(
+                  $scope.story, $scope.getId(), nextNodeId);
+                _init();
+                _recalculateAvailableNodes();
+              });
+            } else {
+              if (nodeId === $scope.getId()) {
+                AlertsService.addInfoMessage(
+                  'A chapter cannot lead to itself.', 3000);
+                return;
+              }
+              if (nodeId === undefined) {
```

Comment:
This shouldn't be permitted, I think. (What exactly is the "default option"?)

Generally try not to use undefined for calculations. All variables in JavaScript are already undefined by default so this practice is quite prone to errors. Use null if you need a value that means "None", or don't have such a value at all.

Reply:
I needed two placeholders, one as the display text - 'Pick a chapter from this  list' and the other, which when clicked would open the modal to create a new chapter, so I used null and undefined. Should I use some other invalid node ids for these placeholders then?

Reply:
No, don't use invalid node IDs at all. If you need a UI element for "completely new node" then treat that as an entirely separate UI element, although it may look visually similar to the previous ones. Otherwise you start losing conceptual integrity since node IDs are no longer always IDs of nodes.

Reply:
Done, removed null and undefined objects from the availableNodes array.